torchtext/csrc/vectors.cpp

```diff
@@ -2,79 +2,83 @@
+          std::to_string(tokens.size()) + ", size of vectors: " +
+          std::to_string(vectors.size(0)) + ".");
     }
 
-    stovectors_.reserve(tokens.size());
+    stovec_.reserve(tokens.size());
     for (std::size_t i = 0; i < tokens.size(); i++) {
       // tokens should not have any duplicates
-      if (stovectors_.find(tokens[i]) != stovectors_.end()) {
+      if (stovec_.find(tokens[i]) != stovec_.end()) {
         throw std::runtime_error("Duplicate token found in tokens list: " +
                                  tokens[i]);
       }
-      stovectors_[tokens[i]] = vectors_[i];
+      stovec_.insert(std::move(tokens[i]), vectors_.select(0, i));
     }
   }
 
   torch::Tensor __getitem__(const std::string &token) const {
-    if (stovectors_.find(token) != stovectors_.end()) {
-      return stovectors_.at(token);
+    const auto &item = stovec_.find(token);
+    if (item != stovec_.end()) {
+      return item->value();
     }
     return unk_tensor_;
   }
 
   void __setitem__(const std::string &token, const torch::Tensor &vector) {
-    stovectors_[token] = vector;
+    const auto &item = stovec_.find(token);
+    if (item != stovec_.end()) {
+      item->value() = vector;
+    } else {
+      tokens_.push_back(token);
+      vectors_ = torch::cat({vectors_, torch::unsqueeze(vector, /*dim=*/0)},
+                            /*dim=*/0);
+      stovec_.insert(token, vectors_.select(0, stovec_.size()));
+    }
   }
 
-  int64_t __len__() { return stovectors_.size(); }
+  int64_t __len__() { return stovec_.size(); }
 };
 
 // Registers our custom class with torch.
 static auto vectors =
     torch::class_<Vectors>("torchtext", "Vectors")
-        .def(torch::init<std::vector<std::string>, std::vector<torch::Tensor>,
+        .def(torch::init<std::vector<std::string>, torch::Tensor,
                          torch::Tensor>())
         .def("__getitem__", &Vectors::__getitem__)
         .def("__setitem__", &Vectors::__setitem__)
         .def("__len__", &Vectors::__len__)
         .def_pickle(
-            // __getstate__
-            [](const c10::intrusive_ptr<Vectors> &self)
-                -> std::tuple<std::vector<std::string>,
-                              std::vector<torch::Tensor>, torch::Tensor> {
-              std::tuple<std::vector<std::string>, std::vector<torch::Tensor>,
-                         torch::Tensor>
+            // __setstate__
+            [](const c10::intrusive_ptr<Vectors> &self) -> std::tuple<
+                std::vector<std::string>, torch::Tensor, torch::Tensor> {
+              std::tuple<std::vector<std::string>, torch::Tensor, torch::Tensor>
                   states(self->tokens_, self->vectors_, self->unk_tensor_);
               return states;
             },
-            // __setstate__
-            [](std::tuple<std::vector<std::string>, std::vector<torch::Tensor>,
+            // __getstate__
+            [](std::tuple<std::vector<std::string>, torch::Tensor,
```

Comment:
Do we want to support a lookup func for a list of tokens?

Reply:
Do you think this would be helpful? I can add this functionality in a followup PR?

Reply:
Done!

Reply:
Didn't we want to do this in a separate PR? Can you split it out please? The title doesn't indicate it's part of it and there's no design discussion around the name or potential overload mechanisms for existing getitem or maybe even a generic map function.